### PR TITLE
add permission decorator and guard

### PR DESCRIPTION
**Pull request by underwater**
add a decoractor and guard to make checking of permissions in controller more declarative (currently not used anywhere, but basis basic unit test)

### CODE REVIEW

File: apps/api/src/guards/has-permissions.guard.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { Injectable, CanActivate, ExecutionContext, HttpException } from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { StatusCodes, getReasonPhrase } from 'http-status-codes';
+import { HAS_PERMISSION_KEY } from '../decorators/has-permission.decorator';
```
Comment (dtslvr) [inline apps/api/src/guards/has-permissions.guard.ts:4]: `import { HAS_PERMISSION_KEY } from '@ghostfolio/api/decorators/has-permission.decorator';`

File: apps/api/src/guards/has-permissions.guard.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import { hasPermission} from '@ghostfolio/common/permissions';
+
+@Injectable()
+export class HasPermissionsGuard implements CanActivate {
+  constructor(private reflector: Reflector) {}
+  canActivate(context: ExecutionContext): boolean {
```
Comment (dtslvr) [inline apps/api/src/guards/has-permissions.guard.ts:10]: Please add a line break

File: apps/api/src/guards/has-permissions.guard.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    const requiredPermission = this.reflector.get<string>(HAS_PERMISSION_KEY, context.getHandler());
+    if (!requiredPermission) {
```
Comment (dtslvr) [inline apps/api/src/guards/has-permissions.guard.ts:12]: Please add a line break

File: apps/api/src/guards/has-permissions.guard.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      return true; // No specific permissions required
+    }
+    const request = context.switchToHttp().getRequest();
```
Comment (dtslvr) [inline apps/api/src/guards/has-permissions.guard.ts:15]: Please add a line break

File: apps/api/src/guards/has-permissions.guard.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    const user = request.user;
+    if (!user || !hasPermission(user.permissions, requiredPermission)) {
```
Comment (dtslvr) [inline apps/api/src/guards/has-permissions.guard.ts:17]: Please add a line break

File: apps/api/src/guards/has-permissions.guard.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      throw new HttpException(
+        getReasonPhrase(StatusCodes.FORBIDDEN),
+        StatusCodes.FORBIDDEN
+      );
+    }
+    return true;
```
Comment (dtslvr) [inline apps/api/src/guards/has-permissions.guard.ts:23]: Please add a line break

**Review by dtslvr — CHANGES_REQUESTED** [review]

**Comment by dtslvr** [comment]
Thank you for your contribution, @underwater! 😎

It already looks fine, just some minor changes to match the overall project's code style.

**Comment by underwater** [comment]
apologies struggling with the github workflow, is anything still pending?

File: apps/api/src/guards/has-permissions.guard.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  ExecutionContext,
+  HttpException
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { StatusCodes, getReasonPhrase } from 'http-status-codes';
+import { HAS_PERMISSION_KEY } from '@ghostfolio/api/decorators/has-permission.decorator';
+import { hasPermission } from '@ghostfolio/common/permissions';
+
+@Injectable()
+export class HasPermissionsGuard implements CanActivate {
```
Comment (dtslvr) [inline apps/api/src/guards/has-permissions.guard.ts:13]: Can you rename to `HasPermissionGuard` because it checks only for one permission, right?

**Review by dtslvr — CHANGES_REQUESTED** [review]
Hi @underwater

I have tried to add `@HasPermission(permissions.accessAdminControl)` to `@Post('flush')` in `CacheController`, but I wasn't able to set a breakpoint in the `HasPermissionsGuard`. Am I missing something?

While testing I had some ideas, I have added a few more comments.

File: apps/api/src/guards/has-permissions.guard.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  constructor(private reflector: Reflector) {}
+
+  canActivate(context: ExecutionContext): boolean {
```
Comment (dtslvr) [inline apps/api/src/guards/has-permissions.guard.ts:16]: Explicitly add `public`...

File: apps/api/src/guards/has-permissions.guard.ts (hunk 7 above)
Comment (underwater) [reply apps/api/src/guards/has-permissions.guard.ts]: done

File: apps/api/src/guards/has-permissions.guard.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    const requiredPermission = this.reflector.get<string>(
+      HAS_PERMISSION_KEY,
+      context.getHandler()
+    );
+
+    if (!requiredPermission) {
+      return true; // No specific permissions required
+    }
+
+    const request = context.switchToHttp().getRequest();
```
[the remark this answers is not in the document]
Comment (underwater) [reply apps/api/src/guards/has-permissions.guard.ts]: done

File: apps/api/src/guards/has-permissions.guard.spec.ts
[the remark this answers is not in the document]
Comment (underwater) [reply apps/api/src/guards/has-permissions.guard.spec.ts:1]: agree

File: apps/api/src/decorators/has-permission.decorator.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
+import { SetMetadata } from '@nestjs/common';
+export const HAS_PERMISSION_KEY = 'has_permission';
+export const HasPermission = (permission: string) =>
```
[the remark this answers is not in the document]
Comment (underwater) [reply apps/api/src/decorators/has-permission.decorator.ts]: done

File: apps/api/src/guards/has-permissions.guard.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  HttpException
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { StatusCodes, getReasonPhrase } from 'http-status-codes';
+import { HAS_PERMISSION_KEY } from '@ghostfolio/api/decorators/has-permission.decorator';
+import { hasPermission } from '@ghostfolio/common/permissions';
+
+@Injectable()
+export class HasPermissionsGuard implements CanActivate {
+  constructor(private reflector: Reflector) {}
```
[the remark this answers is not in the document]
Comment (underwater) [reply apps/api/src/guards/has-permissions.guard.ts]: done

File: apps/api/src/guards/has-permissions.guard.ts (hunk 8 above)
Comment (underwater) [reply apps/api/src/guards/has-permissions.guard.ts]: done

**Comment by underwater** [comment]
> Hi @underwater
> 
> I have tried to add `@HasPermission(permissions.accessAdminControl)` to `@Post('flush')` in `CacheController`, but I wasn't able to set a breakpoint in the `HasPermissionsGuard`. Am I missing something?
> 
> While testing I had some ideas, I have added a few more comments.

I think with the current implementation we'd need to also do

````
  @UseGuards(HasPermissionGuard)  // can be on the controller or handler
  @HasPermission('somePermissionKey') // on the handler
```

I agree this isn't ideal,

File: apps/api/src/guards/has-permission.guard.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  ExecutionContext,
+  HttpException
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { StatusCodes, getReasonPhrase } from 'http-status-codes';
+import { HAS_PERMISSION_KEY } from '@ghostfolio/api/decorators/has-permission.decorator';
+import { hasPermission } from '@ghostfolio/common/permissions';
+
+@Injectable()
+export class HasPermissionsGuard implements CanActivate {
```
Comment (dtslvr) [inline apps/api/src/guards/has-permission.guard.ts:13]: Unfortunately this has been lost: `HasPermissionGuard`

File: apps/api/src/guards/has-permission.guard.spec.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { HasPermissionsGuard as HasPermissionGuard } from './has-permission.guard';
```
Comment (dtslvr) [inline apps/api/src/guards/has-permission.guard.spec.ts:2]: `import { HasPermissionGuard } from './has-permission.guard';`

**Review by dtslvr — CHANGES_REQUESTED** [review]
Thank you very much for the changes and your explanation. I think we can start with `@UseGuards(HasPermissionGuard)` for now.

File: apps/api/src/guards/has-permission.guard.spec.ts (hunk 13 above)
Comment (underwater) [reply apps/api/src/guards/has-permission.guard.spec.ts]: done

**Review by dtslvr — APPROVED** [review]

**Pull request merged**